atomic_reactor/koji_util.py

```diff
@@ -271,3 +284,338 @@ def get_koji_module_build(session, module_spec):
+                'version': docker_version,
+            },
+        ],
+        'extra': {
+            'osbs': {
+                'build_id': build_id,
+                'builder_image_id': get_builder_image_id(build_id, osbs),
+            }
+        },
+        'components': [],
+    }
+    if rpms:
+        buildroot['components'] = get_rpms()
+
+    return buildroot
+
+
+def get_image_output(workflow, image_id, arch):
+    """
+    Create the output for the image
+
+    This is the Koji Content Generator metadata, along with the
+    'docker save' output to upload.
+
+    :return: tuple, (metadata dict, Output instance)
+
+    """
+    saved_image = workflow.exported_image_sequence[-1].get('path')
+    image_name = get_image_upload_filename(workflow.exported_image_sequence[-1],
+                                           image_id, arch)
+
+    metadata = get_output_metadata(saved_image, image_name)
+    output = Output(file=open(saved_image), metadata=metadata)
+
+    return metadata, output
+
+
+def get_image_components(workflow):
+    """
+    Re-package the output of the rpmqa plugin into the format required
+    for the metadata.
+    """
+    output = workflow.image_components
+    if output is None:
+        logger.error("%s plugin did not run!", PostBuildRPMqaPlugin.key)
+        output = []
+
+    return output
+
+
+def get_output(workflow, buildroot_id, pullspec, platform, source_build=False, logs=None):
+    """
+    Build the 'output' section of the metadata.
+    :param buildroot_id: str, buildroot_id
+    :param pullspec: ImageName
+    :param platform: str, output platform
+    :param source_build: bool, is source_build ?
+    :param logs: list, of Output logs
+    :return: tuple, list of Output instances, and extra Output file
+    """
+    def add_buildroot_id(output):
+        logfile, metadata = output
+        metadata.update({'buildroot_id': buildroot_id})
+        return Output(file=logfile, metadata=metadata)
+
+    def add_log_type(output, arch):
+        logfile, metadata = output
+        metadata.update({'type': 'log', 'arch': arch})
+        return Output(file=logfile, metadata=metadata)
+
+    extra_output_file = None
+    output_files = []
+
+    arch = os.uname()[4]
+
+    if source_build:
+        workflow.builder.tasker.pull_image(pullspec)
+        defer_removal(workflow, pullspec)
+        image_inspect = workflow.builder.tasker.inspect_image(pullspec)
+        image_id = image_inspect['Id']
+
+        history = workflow.builder.tasker.get_image_history(image_id)
+        diff_ids = image_inspect[INSPECT_ROOTFS].get(INSPECT_ROOTFS_LAYERS)
+
+        # diff_ids is ordered oldest first
+        # history is ordered newest first
+        # We want layer_sizes to be ordered oldest first
+        layer_sizes = []
+        if diff_ids:
+            layer_sizes = [{"diff_id": diff_id, "size": layer['Size']}
+                           for (diff_id, layer) in zip(diff_ids, reversed(history))]
```

Comment:
If you used patch I suggested, it would be only oneline change. `zip` with empty iterable return empty list this this list comprehension would result into empty list without need for extra ifs and new definitions.

Reply:
ok